File: leecode/code/84/84.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int num = heights.length;
        int max = 0;
        for (int i = 0; i < num; i++) {
            int left = i - 1;
            int right = i + 1;
            while (left >= 0 && heights[left] >= heights[i]) {
                left--;
            }
            while (right < num && heights[right] >= heights[i]) {
                right++;
            }
            max = Math.max(max, heights[i] * (right - left - 1));
        }
        return max;
    }
}

File: leecode/code/84/84.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        res = 0
        rec = list()
        n = len(heights)
        left, right = [0] * n, [n] * n
        for i in range(0, len(heights)):
            while len(rec) > 0 and heights[rec[-1]] > heights[i]:
                k = rec.pop()
                right[k] = i
            left[i] = rec[-1] if rec else -1
            rec.append(i)


        for i in range(0, len(heights)):
            temp = heights[i] * (right[i] - left[i] - 1)
            res = max(res, temp)


        return res